crates/valence_chat/src/lib.rs

```diff
@@ -0,0 +1,632 @@
+}
+
+impl MessageLink {
+    fn update_hash(&self, hasher: &mut impl Digest) {
+        hasher.update(self.sender.into_bytes());
+        hasher.update(self.session_id.into_bytes());
+        hasher.update(self.index.to_be_bytes());
+    }
+}
+
+#[derive(Clone, Debug)]
+struct MessageSignatureStorage {
+    signatures: [Option<[u8; 256]>; 128],
+    indices: FxHashMap<[u8; 256], i32>,
+}
+
+impl Default for MessageSignatureStorage {
+    fn default() -> Self {
+        Self {
+            signatures: [None; 128],
+            indices: FxHashMap::default(),
+        }
+    }
+}
+
+impl MessageSignatureStorage {
+    fn new() -> Self {
+        Self::default()
+    }
+
+    /// Get the index of the `signature` in the storage if it exists.
+    fn index_of(&self, signature: &[u8; 256]) -> Option<i32> {
+        self.indices.get(signature).copied()
+    }
+
+    /// Update the signature storage according to `last_seen` while adding
+    /// `signature` to the storage.
+    ///
+    /// Warning: this consumes `last_seen`.
+    fn add(&mut self, last_seen: &mut VecDeque<[u8; 256]>, signature: &[u8; 256]) {
+        last_seen.push_back(*signature);
+        let mut sig_set = FxHashSet::default();
+        for sig in last_seen.iter() {
+            sig_set.insert(*sig);
+        }
+        for i in 0..128 {
+            if last_seen.is_empty() {
+                return;
+            }
+            // Remove old message
+            let message_sig_data = self.signatures[i];
+            // Add previously seen message
+            self.signatures[i] = last_seen.pop_back();
+            if let Some(data) = self.signatures[i] {
+                self.indices.insert(data, i as i32);
+            }
+            // Reinsert old message if it is not already in last_seen
+            if let Some(data) = message_sig_data {
+                self.indices.remove(&data);
+                if sig_set.insert(data) {
+                    last_seen.push_front(data);
+                }
+            }
+        }
+    }
+}
+
+pub struct SecureChatPlugin;
+
+impl Plugin for SecureChatPlugin {
+    fn build(&self, app: &mut bevy_app::App) {
+        let mojang_pub_key = RsaPublicKey::from_public_key_der(MOJANG_KEY_DATA)
+            .expect("Error creating Mojang public key");
+
+        app.add_plugin(ChatTypePlugin)
+            .insert_resource(MojangServicesState::new(mojang_pub_key))
+            .add_systems(
+                (
+                    init_chat_states,
+                    handle_session_events
+                        .after(init_chat_states)
+                        .before(handle_message_events),
+                    handle_message_acknowledgement
+                        .after(init_chat_states)
+                        .before(handle_message_events),
+                    handle_message_events.after(init_chat_states),
+                )
+                    .in_base_set(CoreSet::PostUpdate)
+                    .before(FlushPacketsSet),
```

Comment:
A couple issues with this schedule:
- The `handle_*` systems need to be scheduled in `EventLoopSchedule` in the `EventLoopSet::PreUpdate` base set so that packets are handled and effects are applied ASAP.
- `init_chat_states` should happen in `CoreSet::PreUpdate` _after_ `SpawnClientsSet` and _before_ `RunEventLoopSet`.